WordPress/gutenberg
# Tag Processor: Fix a problem backing up too far after updating HTML

## What

Fixes a bug in the HTML Tag Processor when updating a document when the internal pointer is at the first tag in the document.

## Why?

The Tag Processor shouldn't break code relying on it.

## How?

A defect introduced in #46018 led to the tag processor backing up one index too far after flushing its queued changes on a document.

For most operations this didn't cause any harm because when immediately moving forward after an update, the `next_tag()` returned to the same spot: it was backing up to one position before the current tag instead of at the start of the current tag.

Unfortunately, when the current tag was the first in the document this would lead the processor to rewind to position `-1`, right before the start of the document, and lead to errors with `strpos()` when it received out-of-bounds indices.

In this fix we're correcting the adjustment for the HTML tag's `<` and documenting the math in the file so that it's clearer why it's there and providing guidance should another fix be necessary.

Props to @anton-vlasenko for finding this bug.

## Testing Instructions

Run the PHP unit test suite with PHP <= 7.0.
Tests should start failing in `trunk` but should pass in this branch.

Temporarily merging in `fix/parity-with-cores-php-ci-jobs` in this PR to run such tests in CI.
